### ISSUE TITLE
More RESTful HTTP handling

### ISSUE THREAD

**Issue by DanMan** [HaxeFoundation/haxe#8614]
Looking at the docs for `haxe.Http` (which is really HttpRequest) and `php.Web` right now, I think there's a bias towards GET and POST requests. Probably because REST wasn't a well known concept when those were created. Some examples:

* You can _send_ a `Http.customRequest` but I don't see how you would handle _receiving_ a new [PUT via PHP](https://www.php.net/manual/en/features.file-upload.put-method.php) or really anything but a GET and POST request.

* There's the unfortunately named `php.Web.getPostData` which actually expects form-encoded data, so it's really `php.Web.getFormData`. It's a nice helper method, but there's nothing inherent to a POST request that wouldn't allow you to send any kind of content, like XML or JSON.

* `haxe.web.Request.getParams` returns "POST parameters", which probably again assumes form-urlencoded data has been sent.

Is it an option to change/expand any of that at this point or is it too late? As is, it looks like Haxe doesn't lend itself to writing a RESTful webservice, at least not with PHP as target. I know PHP isn't "hip" anymore, but I thought I should ask. Maybe I could even help, if there's interest, but I'm not sure if I'm qualified.

**Comment by Aurel300** [comment]
New system APIs, including HTTP functionality, are planned for 4.1. These will be based on Node.js APIs, which are more featureful in terms of methods, data encoding, etc.

**Comment by RealyUniqueName** [comment]
`php.Web` was introduced to be easily swapped with `neko.Web` back in the days when neko was the only server-side target of Haxe. 
Nowadays you better use php-specific functions to handle incoming HTTP requests (until new sys APIs are relased as @Aurel300 said).
E.g. here is [Example #1 from the linked page of php manual](https://www.php.net/manual/en/features.file-upload.put-method.php) translated to Haxe:
```haxe
import php.Global.*;

class Main {
	static function main() {
		/* PUT data comes in on the stdin stream */
		var putdata = fopen("php://input", "r");

		/* Open a file for writing */
		var fp = fopen("myputfile.ext", "w");

		/* Read the data 1 KB at a time and write to the file */
		var data;
		while (data = fread(putdata, 1024))
			fwrite(fp, data);

		/* Close the streams */
		fclose(fp);
		fclose(putdata);
	}
}
```
As you can see it's a straightforward almost one-to-one copy of the php code.
Compile it like this: `haxe -main Main -php bin` if you're using Haxe 4, or add `-D php7` for Haxe 3.4

**Comment by DanMan** [comment]
Interesting, thanks for the hint. `php.Global` is a Haxe 4 thing it seems. In any case, this naturally reduces code re-use, which is one of Haxe's big features. I'd at least expect the basic HTTP stuff to work across both the JS and PHP target. Otherwise I see limited use for Haxe in a webdev context. Those two probably make up 97% of the web after all. Or maybe I haven't written enough Haxe yet to see where it matters the most…

**Comment by RealyUniqueName** [comment]
> php.Global is a Haxe 4 thing it seems.

No, it is also available in Haxe 3.4.7 with `-D php7` flag. But in Haxe 4 it contains more externs for global functions of php. You can use those functions without `php.Global` like this: `var fp = untyped __php__("fopen({0}, {1})", "myputfile.ext", "w");`

**Comment by lublak** [comment]
I'm interested in this.
Is the new API for this http request the async apis coming at some point?
https://github.com/HaxeFoundation/haxe/pull/9111
Is there any estimate when the full implementation will be ready? Since the development is already planned for several years.

**Comment by RealyUniqueName** [comment]
No, unfortunately. It's a heavy task, but we're working on it.
And the latest version of that API does not include any http-related parts.

**Comment by lublak** [comment]
@RealyUniqueName okay :) thank you for the status update. So it's the related githup pull request, so I can track the updates on it.
